internal/rpc/clients.go

```diff
@@ -0,0 +1,155 @@
+	PrivateKeyData    []byte
+}
+
+func (p *ClientParams) usingTLS() bool {
+	return len(p.PublicCertificate) > 0
+}
+
+// GRPCClientFromConfig creates a gRPC client connection from a configuration.
+func GRPCClientFromConfig(cfg config.View, prefix string) (*grpc.ClientConn, error) {
+	hostname := cfg.GetString(prefix + ".hostname")
+	portNumber := cfg.GetInt(prefix + ".grpcport")
+
+	// If TLS support is not enabled in the config, return an insecure grpc client.
+	if !cfg.GetBool("tls.enabled") {
+		return GRPCClientFromParams(&ClientParams{
+			Hostname: hostname,
+			Port:     portNumber,
+		})
+	}
+
+	// If TLS support is enabled, read and return a grpc TLS client.
+	publicKeyPath := cfg.GetString("tls.publicKeyPath")
+
+	publicKeyBytes, err := ioutil.ReadFile(publicKeyPath)
+	if err != nil {
+		clientLogger.WithError(err).Fatalf("cannot read public key from file %s", publicKeyPath)
+		return nil, err
+	}
+
+	return GRPCClientFromParams(&ClientParams{
+		Hostname:          hostname,
+		Port:              portNumber,
+		PublicCertificate: publicKeyBytes,
+	})
+}
+
+// GRPCClientFromParams creates a gRPC client connection from the parameters.
+func GRPCClientFromParams(params *ClientParams) (*grpc.ClientConn, error) {
+	address := fmt.Sprintf("%s:%d", params.Hostname, params.Port)
+	if params.usingTLS() {
+		creds, err := clientCredentialsFromFileData(params.PublicCertificate, address)
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+		return grpc.Dial(address, grpc.WithTransportCredentials(creds))
+	}
+	return grpc.Dial(address, grpc.WithInsecure())
+}
+
+// HTTPClientFromConfig creates a HTTP client from from a configuration.
+func HTTPClientFromConfig(cfg config.View, prefix string) (*http.Client, string, error) {
+	hostname := cfg.GetString(prefix + ".hostname")
+	portNumber := cfg.GetInt(prefix + ".httpport")
+
+	// If TLS support is not enabled in the config, return an insecure grpc client
+	if !cfg.GetBool("tls.enabled") {
+		return HTTPClientFromParams(&ClientParams{
+			Hostname: hostname,
+			Port:     portNumber,
+		})
+	}
+
+	// Read and return an HTTPS client
+	publicKeyPath := cfg.GetString("tls.publicKeyPath")
+	privateKeyPath := cfg.GetString("tls.privateKeyPath")
+
+	publicKeyBytes, err := ioutil.ReadFile(publicKeyPath)
+	if err != nil {
+		clientLogger.WithError(err).Fatalf("cannot read public key from file %s", publicKeyPath)
+		return nil, "", err
+	}
+	privateKeyBytes, err := ioutil.ReadFile(privateKeyPath)
+	if err != nil {
+		clientLogger.WithError(err).Fatalf("cannot read private key from file %s", privateKeyPath)
+		return nil, "", err
+	}
+
+	return HTTPClientFromParams(&ClientParams{
+		Hostname:          hostname,
+		Port:              portNumber,
+		PublicCertificate: publicKeyBytes,
+		PrivateKeyData:    privateKeyBytes,
+	})
+}
+
+// HTTPClientFromParams creates a HTTP client from the parameters.
+func HTTPClientFromParams(params *ClientParams) (*http.Client, string, error) {
```

Comment:
Should this be working with a https://golang.org/pkg/net/url/#URL instead of a raw string?